Test expected failing or unhandled cases

ijpulidos opened the pull request:
This set of changes is aimed at writing test cases that are expected to fail in the protein mutation protocol, either because of limitations of the methodology or because these are simply cases that we don't support to date.

Solves #80

ijpulidos commented:
As usual, just bear in mind that these tests will fail as of today. We will be implementing the protocol to comply to these tests eventually. Some tweaks might be needed when that happens, but the idea is that the overall shape of these tests stay the same when we implement the protein mutation protocol.

IAlibay commented:
> is there a reason generally why we want to avoid

> ring growing

Handling ring breaking is something that the current hybrid topology cannot handle properly - it's a fundamental issue in the way in which we deal with bond scaling. There are some ways around this, but they would require a rather large rewrite of the hybrid topology approach and generally isn't something we want to invest a lot of our time on.

> double charge changes

This is related to https://github.com/OpenFreeEnergy/openfe/issues/660

The main answer here is "the tooling doesn't allow it now, we could do it, but it's not the best use of our resources".

IAlibay commented on feflow/utils/exceptions.py:
```diff
@@ -0,0 +1,15 @@
+    """Base exception for OpenFE protocols"""
+
+
+class MethodConstraintError(ProtocolError):
```
This name is a bit unintuitive imho, it makes me think of "constraints".

IAlibay commented on feflow/utils/exceptions.py:
```diff
@@ -0,0 +1,15 @@
+"""Module with custom exceptions that can be useful for including in protocols"""
+
+
+class ProtocolError(Exception):
```
If you want to create a base error type, this really should be in gufe

IAlibay commented on feflow/utils/exceptions.py:
```diff
@@ -0,0 +1,15 @@
+    that does not support."""
+
+
+class NotSupportedError(ProtocolError):
```
This name is somewhat unspecific for something that inherits from something Protocol specific, could there be a better name?

IAlibay reviewed (commented):
I'm not super convinced on the error names, might be something we need to workshop in the longer run.

jthorton commented on feflow/tests/test_protein_mutation.py:
```diff
@@ -476,3 +525,71 @@ def test_charge_changing_convergence(
+        protocol = ProteinMutationProtocol(settings=settings)
+
+        # Expect an error when trying to create the DAG with this invalid transformation
+        with pytest.raises(MethodConstraintError, match="proline.*not supported"):
```
I am unsure if it's a valid use case, but do you intend to stop all mutations involving proline? So, a proline-to-proline mutation would not be allowed in very different conformations as well?

ijpulidos commented:
I renamed some of the custom errors and rewrote some of the strings, I think  we could probably reiterate on these if needed in the future. I'll go ahead and merge this.
